database/sql/V013__create_presignatures.sql

```diff
@@ -0,0 +1,12 @@
+-- PreSignature events from the smart contract.
+CREATE TABLE presignatures (
+    block_number bigint NOT NULL,
+    log_index bigint NOT NULL,
+    owner bytea NOT NULL,
+    order_uid bytea NOT NULL,
+    signed boolean NOT NULL,
+    PRIMARY KEY (block_number, log_index)
```

Comment:
Since presignatures are stored in `mapping(orderUid => bool)`, doesn't it make sense for the primary key to be the order UID in the table?

I guess it depends on what the expectations are for:
- User create presignature for order X
- User remove presignature for order X

In the above scenario, should the presignature row get removed? Should there be a new entry in it?

Reply:
Actually, now that I think about it a bit more, I think there would be an issue if we use the `orderUid` as the primary key, because we emit `PreSignature` events even if the value doesn't change. So
- User sets presignature for order X on block N
- User sets presignature again, for absolutely no reason, on block M with Tx T nonce O
- User has another Tx T' with nonce O in the mempool
- Reorg to block M-1 happens
  - presignature row gets deleted, even if the signature from block N is still there
- T' gets included in the canonical chain instead of T
- No presignature.

Reply:
Other tables that store on chain events have this primary key because it uniquely identifies any event. This is also needed to efficiently query all events in some block range which is used for the reord which deletes all events in the last X blocks.
That doesn't mean it has to be a primary key here but it has to be at least an index (primary keys are always indexes).

Reply:
This table keeps all presign events as independent rows. To my understanding we want to keep all events since it might be necessary to roll back in case of a reorg. This extra information is indeed not needed, the alternative could be storing everything in memory for the first `MAX_REORG_BLOCK_COUNT` and only then finalize the presignatures into the database after waiting that number of blocks. I think the latter solution is fairly more complex (requires reworking the event abstraction) but definitely nicer.
I went for the current implementation because it looked significantly easier, and it can be made (imho) comparably as efficient as the other by pruning unnecessary events in a maintenance step.